cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -445,29 +478,46 @@ class GcsFileSystem::Impl {
   }
 
   Status DeleteFile(const GcsPath& p) {
-    if (!p.object.empty() && p.object.back() == '/') {
-      return Status::IOError("The given path (" + p.full_path +
-                             ") is a directory, use DeleteDir");
+    if (!p.object.empty()) {
+      auto stat = client_.GetObjectMetadata(p.bucket, p.object);
+      if (!stat) return internal::ToArrowStatus(stat.status());
+      if (IsDirectory(*stat)) {
+        return Status::IOError("The given path '", p.full_path,
+                               "' is a directory, use DeleteDir");
+      }
     }
     return internal::ToArrowStatus(client_.DeleteObject(p.bucket, p.object));
   }
 
   Status Move(const GcsPath& src, const GcsPath& dest) {
-    if (src.full_path.empty() || src.object.empty() ||
-        src.object.back() == internal::kSep) {
+    if (src == dest) return {};
```

Comment:
nit: return Status::OK() is more idiomatic for the code base

Reply:
Fixed here and elsewhere.